linux/src/threadpool_linux.c

```diff
@@ -103,7 +117,10 @@ static void on_timer_callback(sigval_t timer_data)
     }
     else
     {
-        timer_instance->work_function(timer_instance->work_function_ctx);
+        if (interlocked_add(&timer_instance->timer_status, 0) == TIMER_ENABLED)
+        {
+            timer_instance->work_function(timer_instance->work_function_ctx);
+        }
     }
 }
```

Comment:
Do we have any sort of unique identifier to match in the `timer_data` with the timer start?

Currently I think your solution avoids a segfault, but it's not correct in all cases.

Imagine you have 63 timers set already.

Now:
1. start a timer (A) with a 1 second period
2. some time later stop the timer (A), which will set the state to DISABLED
3. start a timer (B) with a 1 hour initial period

It's entirely possible that timer A fires after you delete it (hence your fix).

But now the handler for timer B will see the timer intended for A.

Clearly that's broken if we rely on this for something like a watchdog (we could re-use a timer slot and immediately fire the watchdog).

I think you could even catch this in a chaos test (fill up the 63 slots then have a thread that alternates between starting a very short timer and starting a timer with a long enough time that it will never fire and assert if it does).

Reply:
Possibly could be solved by having some extra indirection/lookup.

If instead of this:

```
TIMER_INSTANCE* timer_instance = timer_data.sival_ptr;
```

we had some intermediate:

```
TIMER_IDENTIFIER timer_identifier = timer_data.sival_ptr;
TIMER_INSTANCE* timer_instance = some_lookup_function(timer_identifier);
```

But I'm looking at sigval now and I see we only have the one `void*` or `int` and we can't have another context.

I think we really need 2 fields in the callback to do what I'm thinking of, because my "some_lookup_function" likely needs to be "TIMER_INSTANCE* some_lookup_function(THREADPOOL* threadpool, TIMER_IDENTIFIER timer_identifier)" or you would have to have a global array in that lookup function.

Reply:
The best way to solve this would be for the calling function in the callback to have the unique ID.  This would make sure that the callback is always set for you, but it does put undue work on the calling function.  I added a chaos test for this but I haven't gotten it to fail yet (I probably need to tweak it a little).  I don't know if this is an actual problem or an in-theory problem.  I'll continue to look for a reliable solution

Reply:
I'm not sure how you would check for a unique ID. You have to store it somewhere, so if you store it in the fixed array, then it's still the same problem.

If the caller stores it in their context I don't see how that helps.

Timer A will have context A (with unique ID 1)
Timer B will have context B (with unique ID 2)

But say timer slot 0 is re-used.
First it gets context A (id 1)
If we re-use slot 0 with context B and it fires due to the original registration, then it will just call the callback for B with context B and unique ID 2.

So even if the timer is due to A, it will be handled by B and look exactly like it "should" be for for B.

Maybe a simpler work-around for this tiny race is to have some timestamp or something so we block before re-using a timer slot (e..g don't re-use the slot until 1-5 seconds or something has elapsed or something).

Reply:
I think I might have a solution.  Let me know what you think.